modules/ml/src/main/java/org/apache/ignite/ml/naivebayes/gaussian/GaussianNaiveBayesTrainer.java

```diff
@@ -0,0 +1,297 @@
+    }
+
+    /** {@inheritDoc} */
+    @Override protected boolean checkState(GaussianNaiveBayesModel mdl) {
+        return true;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected <K, V> GaussianNaiveBayesModel updateModel(GaussianNaiveBayesModel mdl,
+        DatasetBuilder<K, V> datasetBuilder, IgniteBiFunction<K, V, Vector> featureExtractor,
+        IgniteBiFunction<K, V, Double> lbExtractor) {
+        assert datasetBuilder != null;
+
+        PartitionDataBuilder<K, V, EmptyContext, LabeledVectorSet<Double, LabeledVector>> partDataBuilder
+            = new LabeledDatasetPartitionDataBuilderOnHeap<>(
+            featureExtractor,
+            lbExtractor
+        );
+
+        try (Dataset<EmptyContext, LabeledVectorSet<Double, LabeledVector>> dataset = datasetBuilder.build(
+            (upstream, upstreamSize) -> new EmptyContext(),
+            partDataBuilder
+        )) {
+            MeanHelper mean = computeMeans(dataset);
+
+            List<Double> sortedFeatures = new ArrayList<>(mean.featureCount.keySet());
+            sortedFeatures.sort(Double::compareTo);
+
+            int labelCount = sortedFeatures.size();
+            int featureCount = mean.featureSum.get(sortedFeatures.get(0)).length;
+
+            double[][] means = new double[labelCount][featureCount];
+            double[][] variances = new double[labelCount][featureCount];
+            double[] classProbabilities = new double[labelCount];
+            double[] labels = new double[labelCount];
+
+            long datasetSize = mean.featureCount.values().stream().mapToInt(i -> i).sum();
+            Map<Double, double[]> meansHolder = new HashMap<>();
+
+            int lbl = 0;
+            for (Double label : sortedFeatures) {
+                int count = mean.featureCount.get(label);
+                double[] tmp = mean.featureSum.get(label);
+
+                for (int i = 0; i < featureCount; i++) {
+                    means[lbl][i] = tmp[i] / count;
+                }
+
+                meansHolder.put(label, means[lbl]);
+                classProbabilities[lbl] = (double)count / datasetSize;
+                labels[lbl] = label;
+                ++lbl;
+            }
+
+            VarianceHelper variance = computeVariance(dataset, meansHolder);
+
+            lbl = 0;
+            for (Double label : sortedFeatures) {
+                int count = mean.featureCount.get(label);
+                double[] tmp = variance.featureDiff.get(label);
+
+                for (int i = 0; i < featureCount; i++) {
+                    variances[lbl][i] = tmp[i] / (count - 1);
+                }
+                ++lbl;
+            }
+            if(equiprobableClasses){
+                int k = classProbabilities.length;
+                for (int i = 0; i < k; i++) {
+                    classProbabilities[i] = 1./k;
+                }
+            }
+            if (priorProbabilities != null){
```

Comment:
did you forget "else" ?
As I see "priorProbabilities != null" and "equiprobableClasses" is a mutually exclusive conditions
Moreover, this step may be placed in "for (Double label : sortedFeatures) { ... } " cycle above for mean values computing.

Reply:
You are right, these options are mutually exclusive. I put this step inside the loop. Logically it's a good place for it because we have `if-else_if-else` condition. But the loop body became bigger, I'm afraid that it can affect readability.